source/auth/auth.rst

```diff
@@ -1159,6 +1159,199 @@ If AWS authentication fails for any reason, the cache MUST be cleared.
+
+
+`MongoCredential`_ Properties
+`````````````````````````````
+
+username
+    MUST NOT be specified.
+
+source
+    MUST be "$external". Defaults to ``$external``.
+
+password
+    MUST NOT be specified.
+
+mechanism
+    MUST be "MONGODB-OIDC"
+
+mechanism_properties
+    PRINCIPAL_NAME
+        Drivers MUST allow the user to specify a principal name, which
+        is required by the server when more than one Identity Provider
+        is configured.
+    DEVICE_NAME
+        Drivers MUST allow the user to specify a name for the device
+        workflow that is one of "aws", "azure", or "gcp".
+
+
+User Provided Callbacks
+```````````````````````
+
+Drivers MUST allow the user to provide callbacks for token request and
+token refresh.  The token request callback MUST accept the OIDCMechanismServerStep1 structure and return an OIDCRequestTokenResult of
+the form:
+
+.. code:: idl
+
+  OIDCRequestTokenResult:
+      description: "The result of a token request"
+      strict: false
+      fields:
+          accessToken:
+              description: "The OIDC access token"
+              type: string
+          expiresInSeconds:
+              description: "The expiration time in seconds from the current time"
+              type: int
+              optional: true
+          refreshToken:
+              description: "The OIDC refresh token"
+              type: str
+              optional: true
+
+.. code:: typescript
+
+    function onOIDCRequestToken(serverInfo: OIDCMechanismServerStep1): OIDCRequestTokenResult
+
+The optional token refresh callback will accept the IDP information as
+well as the cached OIDCRequestTokenResult and return a new OIDCRequestTokenResult.
+
+
+.. code:: typescript
+
+    function onOIDCRefreshToken(serverInfo: OIDCMechanismServerStep1, tokenResult: OIDCRequestTokenResult): OIDCRequestTokenResult
+
+If no callbacks are given, the driver MUST enforce that a DEVICE_NAME
+mechanism_properties is set and one of ("aws", "azure", or "gcp").
+The callback mechanism can be used to support both Authentication
+Code Workflows or Device workflows that are not explicitly implemented
+by drivers.
+
+
+Supported Device Workflows
+``````````````````````````
+
+Drives MUST support device workflows for "aws", "azure", and "gcp", given
+by the DEVICE_NAME mechanism property.  In all cases the acquired token
+will be given as the ``jwt`` argument in the second client step of the
+SASL exchange.
+
+AWS
+___
+
+When the DEVICE_NAME mechanism property is set to "aws", the driver MUST
+attempt to read the value given by the ``AWS_WEB_IDENTITY_TOKEN_FILE`` and
```

Comment:
Can we use the full aws workflow as in aws kms (similar to below approach for gcp and azure)?

Reply:
based on .net aws sdk, in this case we will need to look at `AWS_ROLE_ARN` too (otherwise web-identity workflow will fail) as I know we discsussed we don't want doing it, but if we put this aside, will accessToken provided by let's say ec2/ecs envs work here?